File: CSharpCodeBase/components/thing/thingregenerationcomponent.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class thingregenerationcomponent :component{
 var PercentLifeRegenerationString = "PercentLifeRegeneration";
 var TimeUpdateLifeRegenerationString = "TimeUpdateLifeRegeneration";
 var LifeRegenerationNextStateString = "LifeRegenerationNextState";
 public void init(self, entity);
   Component.init(self, entity);
   this.entity.regeneration = self;
       
   this.percentRegeneration = this.entity.config:Get("percentRegeneration")  ||  10;
   this.timeStep = this.entity.config:Get("timeStepRegeneration")  ||  25;
   self:SetPhases(this.entity.config:Get("regenerationPhases")  ||  "none");
       
   this.nextRegenerationTimeCheckPoint = this.timeStep + GameController.gameTime;
   this.healthRegenerationValue = (this.entity.health.maxAmount/100)*this.percentRegeneration      ;
   this.isStarted = false;
   
   self:Start();
 }})
 public void SetPhases(value){
    this.regenerationTime = StringUtils.split(value, ",");
 }
 public void Start(){
   this.nextRegenerationTimeCheckPoint = this.timeStep + GameController.gameTime;
   this.isStarted = true;
 }
 public void Stop(){
   this.isStarted = false;
 }
 public void Update(){
   if(this.isStarted  ){ 
     self:PassiveRegeneration();
   }
 }
 public void PassiveRegeneration(){
   if(GameController.gameTime >= this.nextRegenerationTimeCheckPoint  ){ 
    for(key, value in pairs(this.regenerationTime) ){
       if(GameController.daytime.currentPhase == GameController.daytime[value]  ||  value == "none"  ){  
         this.nextRegenerationTimeCheckPoint = this.timeStep + GameController.gameTime;
         this.entity:OnRegenerate(this.healthRegenerationValue);
         return;
       }
     }
   }
 }
 }}

File: CSharpCodeBase/components/thing/thingregenerationcomponent.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'

local PercentLifeRegenerationString = 'PercentLifeRegeneration'
local TimeUpdateLifeRegenerationString = 'TimeUpdateLifeRegeneration'
local LifeRegenerationNextStateString = 'LifeRegenerationNextState'

local ThingRegenerationComponent = Class({__includes=Component, init=function(self, entity)
  Component.init(self, entity)
  self.entity.regeneration = self
      
  self.percentRegeneration = self.entity.config:Get('percentRegeneration') or 10
  self.timeStep = self.entity.config:Get('timeStepRegeneration') or 25
  self:SetPhases(self.entity.config:Get('regenerationPhases') or 'none')
      
  self.nextRegenerationTimeCheckPoint = self.timeStep + GameController.gameTime
  self.healthRegenerationValue = (self.entity.health.maxAmount/100)*self.percentRegeneration      
  self.isStarted = false
  
  self:Start()
end})

function ThingRegenerationComponent:SetPhases(value)
   self.regenerationTime = StringUtils.split(value, ',')
end

function ThingRegenerationComponent:Start()
  self.nextRegenerationTimeCheckPoint = self.timeStep + GameController.gameTime
  self.isStarted = true
end

function ThingRegenerationComponent:Stop()
  self.isStarted = false
end

function ThingRegenerationComponent:Update()
  if self.isStarted then
    self:PassiveRegeneration()
  end
end

function ThingRegenerationComponent:PassiveRegeneration()
  if GameController.gameTime >= self.nextRegenerationTimeCheckPoint then
   for key, value in pairs(self.regenerationTime) do
      if GameController.daytime.currentPhase == GameController.daytime[value] or value == 'none' then 
        self.nextRegenerationTimeCheckPoint = self.timeStep + GameController.gameTime
        self.entity:OnRegenerate(self.healthRegenerationValue)
        return
      end
    end
  end
end

return ThingRegenerationComponent